databases/nmdc/database.go

```diff
@@ -0,0 +1,992 @@
+	request, err := http.NewRequest(http.MethodPost, resource, strings.NewReader(data.Encode()))
+	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	request.Header.Set("Accept", "application/json")
+
+	var client http.Client
+	response, err := client.Do(request)
+	if err != nil {
+		return auth, err
+	}
+
+	switch response.StatusCode {
+	case 200, 201, 204:
+		defer response.Body.Close()
+		var data []byte
+		data, err = io.ReadAll(response.Body)
+		if err != nil {
+			return auth, err
+		}
+		type accessTokenResponse struct {
+			Token   string `json:"access_token"`
+			Type    string `json:"token_type"`
+			Expires struct {
+				Days    int `json:"days"`
+				Hours   int `json:"hours"`
+				Minutes int `json:"minutes"`
+			} `json:"expires"`
+		}
+		var tokenResponse accessTokenResponse
+		err = json.Unmarshal(data, &tokenResponse)
+		if err != nil {
+			return auth, err
+		}
+		// calculating the time of expiry, subtracting 1 minute for "slop"
+		duration := time.Duration(24*tokenResponse.Expires.Days+tokenResponse.Expires.Hours)*time.Hour +
+			time.Duration(tokenResponse.Expires.Minutes-1)*time.Minute
+		return authorization{
+			Credential:     credential,
+			Token:          tokenResponse.Token,
+			Type:           tokenResponse.Type,
+			Expires:        true,
+			ExpirationTime: time.Now().Add(duration),
+		}, err
+	case 503:
+		return auth, &databases.UnavailableError{
+			Database: "nmdc",
+		}
+	default:
+		defer response.Body.Close()
+		var data []byte
+		data, _ = io.ReadAll(response.Body)
+		type errorResponse struct {
+			Detail string `json:"detail"`
+		}
+		var errResponse errorResponse
+		err = json.Unmarshal(data, &errResponse)
+		if err != nil {
+			return auth, err
+		}
+		return auth, &databases.UnauthorizedError{
+			Database: "nmdc",
+			User:     credential.User,
+			Message:  errResponse.Detail,
+		}
+	}
+}
+
+// checks our access token for expiration and renews if necessary
+func (db *Database) renewAccessTokenIfExpired() error {
+	var err error
+	if time.Now().After(db.Auth.ExpirationTime) { // token has expired
+		db.Auth, err = getAccessToken(db.Auth.Credential)
+	}
+	return err
+}
+
+// adds an appropriate authorization header to given HTTP request
+func (db Database) addAuthHeader(request *http.Request) {
+	request.Header.Add("Authorization", fmt.Sprintf("Bearer %s", db.Auth.Token))
+}
+
+// performs a GET request on the given resource, returning the resulting
+// response body and/or error
+func (db Database) get(resource string, values url.Values) ([]byte, error) {
+	res, err := url.Parse(baseApiURL)
+	if err != nil {
+		return nil, err
+	}
+	res.Path += resource
+	res.RawQuery = values.Encode()
+	slog.Debug(fmt.Sprintf("GET: %s", res.String()))
+	req, err := http.NewRequest(http.MethodGet, res.String(), http.NoBody)
+	if err != nil {
+		return nil, err
+	}
+	db.addAuthHeader(req)
+	resp, err := db.Client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	switch resp.StatusCode {
+	case 200:
```

Comment:
Again, I'm assuming that you got intel from the nmdc devs on valid responses from this endpoint